lib/context.js

```diff
@@ -0,0 +1,258 @@
+      // use the original `externalKey` if exists to avoid waste storage
+      this.create(val, this.externalKey);
+      return;
+    }
+    throw new Error('this.session can only be set as null or an object.');
+  }
+
+  /**
+   * init session from external store
+   * will be called in the front of session middleware
+   *
+   * @api public
+   */
+
+  * initFromExternal() {
+    debug('init from external');
+    const ctx = this.ctx;
+    const opts = this.opts;
+
+    const externalKey = ctx.cookies.get(opts.key, opts);
+    debug('get external key from cookie %s', externalKey);
+
+    if (!externalKey) {
+      // create a new `externalKey`
+      this.create();
+      return;
+    }
+
+    const json = yield this.store.get(externalKey, ctx);
+    if (!this.valid(json)) {
+      // create a new `externalKey`
+      this.create();
+      return;
+    }
+
+    // create with original `externalKey`
+    this.create(json, externalKey);
+    this.prevHash = util.hash(this.session.toJSON());
+  }
+
+  /**
+   * init session from cookie
+   * @api private
+   */
+
+  initFromCookie() {
+    debug('init from cookie');
+    const ctx = this.ctx;
+    const opts = this.opts;
+
+    const cookie = ctx.cookies.get(opts.key, opts);
+    if (!cookie) {
+      this.create();
+      return;
+    }
+
+    let json;
+    debug('parse %s', cookie);
+    try {
+      json = opts.decode(cookie);
+    } catch (err) {
+      // backwards compatibility:
+      // create a new session if parsing fails.
+      // new Buffer(string, 'base64') does not seem to crash
+      // when `string` is not base64-encoded.
+      // but `JSON.parse(string)` will crash.
+      debug('decode %j error: %s', cookie, err);
+      if (!(err instanceof SyntaxError)) throw err;
+      this.create();
+      return;
+    }
+
+    debug('parsed %j', json);
+
+    if (!this.valid(json)) {
+      this.create();
+      return;
+    }
+
+    // support access `ctx.session` before session middleware
+    this.create(json);
+    this.prevHash = util.hash(this.session.toJSON());
+  }
+
+  /**
+   * verify session(expired or )
+   * @param  {Object} json session object
+   * @return {Boolean} valid
+   * @api private
+   */
+
+  valid(json) {
+    if (!json) return false;
+
+    if (!json._expire || json._expire < Date.now()) {
+      debug('expired session');
+      return false;
+    }
+
+    const valid = this.opts.valid;
+    if (typeof valid === 'function' && !valid(this.ctx, json)) {
+      // valid session value fail, ignore this session
+      debug('invalid session');
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * create a new session and attach to ctx.sess
+   *
+   * @param {Object} [val] session data
+   * @param {String} [externalKey] session external key
+   * @api private
+   */
+
+  create(val, externalKey) {
+    debug('create session with val: %j externalKey: %s', val, externalKey);
+    if (this.opts.store) this.externalKey = externalKey || uid.sync(24);
```

Comment:
yes, can add `options.genid` later, don't want to add too many things in this PR.